## Homework #1 

### Set up the environment

You need to install Python, NumPy, Pandas, Matplotlib and Seaborn. For that, you can the instructions from <a href='https://github.com/alexeygrigorev/mlbookcamp-code/blob/master/course-zoomcamp/01-intro/06-environment.md'>06-environment.md</a>.

In [1]:
import numpy as np
import pandas as pd

### Question 1

What's the version of Pandas that you installed?

You can get the version information using the `__version__` field:

In [2]:
pd.__version__

'1.5.3'

### Getting the data

For this homework, we'll use the California Housing Prices dataset. Download it from <a href='https://raw.githubusercontent.com/alexeygrigorev/datasets/master/housing.csv'>here</a>. 

You can do it with wget:
```bash
wget https://raw.githubusercontent.com/alexeygrigorev/datasets/master/housing.csv
```
Or just open it with your browser and click "Save as...".

Now read it with Pandas.

In [3]:
data = pd.read_csv("https://raw.githubusercontent.com/alexeygrigorev/datasets/master/housing.csv")

### Question 2

How many records are in the dataset?

In [4]:
data.shape[0]   # alternative is len(data.index)

20640

### Question 3

Which columns in the dataset have missing values?

In [5]:
data['total_rooms'].isnull().sum() != 0

False

In [6]:
data['total_bedrooms'].isnull().sum() != 0

True

We can see that `total_bedrooms` column return `True` value, so there are missing values on that column.

Or, we can also see the total number of missing value:

In [7]:
data.isnull().sum()

longitude               0
latitude                0
housing_median_age      0
total_rooms             0
total_bedrooms        207
population              0
households              0
median_income           0
median_house_value      0
ocean_proximity         0
dtype: int64

### Question 4

How many unique values does the `ocean_proximity` column have?

In [8]:
data['ocean_proximity'].nunique()

5

We can see that `ocean_proximity` column have 5 unique value

We can also see each value of this column:

In [9]:
data['ocean_proximity'].unique().tolist()

['NEAR BAY', '<1H OCEAN', 'INLAND', 'NEAR OCEAN', 'ISLAND']

### Question 5

What's the average value of the `median_house_value` for the houses located near the bay?

In [10]:
int(data[data['ocean_proximity'] == 'NEAR BAY']['median_house_value'].mean())

259212

### Question 6

1. Calculate the average of `total_bedrooms` column in the dataset.
2. Use the `fillna` method to fill the missing values in `total_bedrooms` with the mean value from the previous step.
3. Now, calculate the average of `total_bedrooms` again.
4. Has it changed?

Has it changed?

Firstly, we calculate the `old` data for `total_bedrooms` column, we can see that there are `207` columns have missing values

In [11]:
old_total_bedroomss_data = data['total_bedrooms'].isnull().sum()
old_total_bedroomss_data

207

In [12]:
old_average_total_bedrooms = round(data['total_bedrooms'].mean(),3)
print(f"Average for Old Data in total_bedrooms Column = {old_average_total_bedrooms}")

Average for Old Data in total_bedrooms Column = 537.871


We also calculate the missing value of `new` data for `total_bedrooms` column after we fill the missing value columns with the average of old data

We can see that, there are no missing value on the `new` data of `total_bedrooms` column

In [13]:
new_total_bedroomss_data = data['total_bedrooms'].fillna(old_average_total_bedrooms)
new_total_bedroomss_data.isnull().sum()

0

In [14]:
new_average_total_bedrooms = round(new_total_bedroomss_data.mean(),3)
print(f"Average for New Data in total_bedrooms Column = {new_average_total_bedrooms}")

Average for New Data in total_bedrooms Column = 537.871


`No, it left the same` with the 3 digits after the decimal point.

old_average_total_bedrooms = 537.871

new_average_total_bedrooms = 537.871

It should be have impact on the dataset, however because the dataset have the large row, so it only give the little impact. 

### Question 7

1. Select all the options located on islands.
2. Select only columns `housing_median_age`, `total_rooms`, `total_bedrooms`.
3. Get the underlying NumPy array. Let's call it `X`.
4. Compute matrix-matrix multiplication between the transpose of `X` and `X`. To get the transpose, use `X.T`. Let's call the result `XTX`.
5. Compute the inverse of `XTX`.
6. Create an array `y` with values `[950, 1300, 800, 1000, 1300]`.
7. Multiply the inverse of `XTX` with the transpose of `X`, and then multiply the result by `y`. Call the result `w`.
8. What's the value of the last element of `w`?

In [15]:
islands_df = data[data['ocean_proximity'] == 'ISLAND']
islands_df

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
8314    -118.32     33.35                27.0       1675.0           521.0   
8315    -118.33     33.34                52.0       2359.0           591.0   
8316    -118.32     33.33                52.0       2127.0           512.0   
8317    -118.32     33.34                52.0        996.0           264.0   
8318    -118.48     33.43                29.0        716.0           214.0   

      population  households  median_income  median_house_value  \
8314       744.0       331.0         2.1579            450000.0   
8315      1100.0       431.0         2.8333            414700.0   
8316       733.0       288.0         3.3906            300000.0   
8317       341.0       160.0         2.7361            450000.0   
8318       422.0       173.0         2.6042            287500.0   

     ocean_proximity  
8314          ISLAND  
8315          ISLAND  
8316          ISLAND  
8317          ISLAND  
8318          ISLAND

In [16]:
islands_df = islands_df[['housing_median_age', 'total_rooms', 'total_bedrooms']]
islands_df

housing_median_age  total_rooms  total_bedrooms
8314                27.0       1675.0           521.0
8315                52.0       2359.0           591.0
8316                52.0       2127.0           512.0
8317                52.0        996.0           264.0
8318                29.0        716.0           214.0

In [17]:
X = islands_df.to_numpy()
X

array([[  27., 1675.,  521.],
       [  52., 2359.,  591.],
       [  52., 2127.,  512.],
       [  52.,  996.,  264.],
       [  29.,  716.,  214.]])

In [18]:
XTX = X.T @ X
XTX

array([[9.6820000e+03, 3.5105300e+05, 9.1357000e+04],
       [3.5105300e+05, 1.4399307e+07, 3.7720360e+06],
       [9.1357000e+04, 3.7720360e+06, 9.9835800e+05]])

In [19]:
XTX_inv = np.linalg.inv(XTX)
XTX_inv

array([[ 9.19403586e-04, -3.66412216e-05,  5.43072261e-05],
       [-3.66412216e-05,  8.23303633e-06, -2.77534485e-05],
       [ 5.43072261e-05, -2.77534485e-05,  1.00891325e-04]])

In [20]:
y = np.array([950, 1300, 800, 1000, 1300])
y

array([ 950, 1300,  800, 1000, 1300])

In [21]:
w = (XTX_inv @ X.T) @ y
w

array([23.12330961, -1.48124183,  5.69922946])

In [22]:
round(w[-1],4)

5.6992

> **Note**: we just implemented normal equation


$$w = (X^T X)^{-1} X^T y$$